x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/persistence/ResultsPersisterService.java

```diff
@@ -0,0 +1,185 @@
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+
+import static org.elasticsearch.xpack.core.ClientHelper.ML_ORIGIN;
+
+public class ResultsPersisterService {
+    private static final Logger LOGGER = LogManager.getLogger(ResultsPersisterService.class);
+
+    public static final Setting<Integer> PERSIST_RESULTS_MAX_RETRIES = Setting.intSetting(
+        "xpack.ml.persist_results_max_retries",
+        20,
+        0,
+        50,
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope);
+    private static final int MAX_RETRY_SLEEP_MILLIS = (int)Duration.ofMinutes(15).toMillis();
+    private static final int MIN_RETRY_SLEEP_MILLIS = 50;
+    // Having an exponent higher than this causes integer overflow
+    private static final int MAX_RETRY_EXPONENT = 24;
+
+    private final Random random = Randomness.get();
+    private final Client client;
+    private volatile int maxFailureRetries;
+
+    public ResultsPersisterService(Client client, ClusterService clusterService, Settings settings) {
+        this.client = client;
+        this.maxFailureRetries = PERSIST_RESULTS_MAX_RETRIES.get(settings);
+        clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(PERSIST_RESULTS_MAX_RETRIES, this::setMaxFailureRetries);
+    }
+
+    void setMaxFailureRetries(int value) {
+        this.maxFailureRetries = value;
+    }
+
+    public BulkResponse indexWithRetry(String jobId,
+                                       String indexName,
+                                       ToXContent object,
+                                       ToXContent.Params params,
+                                       WriteRequest.RefreshPolicy refreshPolicy,
+                                       String id,
+                                       Supplier<Boolean> shouldRetry,
+                                       Consumer<String> msgHandler) throws IOException {
+        BulkRequest bulkRequest = new BulkRequest().setRefreshPolicy(refreshPolicy);
+        try (XContentBuilder content = object.toXContent(XContentFactory.jsonBuilder(), params)) {
+            bulkRequest.add(new IndexRequest(indexName).id(id).source(content));
+        }
+        return bulkIndexWithRetry(bulkRequest, jobId, shouldRetry, msgHandler);
+    }
+
+    public BulkResponse bulkIndexWithRetry(BulkRequest bulkRequest,
+                                           String jobId,
+                                           Supplier<Boolean> shouldRetry,
+                                           Consumer<String> msgHandler) {
+        return bulkIndexWithRetry(bulkRequest, jobId, shouldRetry, msgHandler, 10);
+    }
+
+    BulkResponse bulkIndexWithRetry(BulkRequest bulkRequest,
+                                    String jobId,
+                                    Supplier<Boolean> shouldRetry,
+                                    Consumer<String> msgHandler,
```

Comment:
could the msgHandler do both: log and audit?

Reply:
Sure, whatever the handler wants, but I do think it is important for results persister to log on its own.